src/Processors/Transforms/PartialSortingTransform.cpp

```diff
@@ -55,33 +81,39 @@ void PartialSortingTransform::transform(Chunk & chunk)
     ColumnRawPtrs block_columns;
     UInt64 rows_num = block.rows();
 
+    /** If we've saved columns from previously blocks we could filter all rows from current block
+      * which are unnecessary for sortBlock(...) because they obviously won't be in the top LIMIT rows.
+      */
     if (!threshold_block_columns.empty())
     {
-        IColumn::Filter filter(rows_num, 0);
         block_columns = extractColumns(block, description);
-        size_t filtered_count = 0;
-
-        for (UInt64 i = 0; i < rows_num; ++i)
-        {
-            if (less(threshold_block_columns, limit - 1, block_columns, i, description))
-            {
-                ++filtered_count;
-                filter[i] = 1;
-            }
-        }
+        IColumn::Filter filter;
+        size_t filtered_count = getFilterMask(block_columns, threshold_block_columns, limit - 1, description, rows_num, filter);
 
         if (filtered_count)
         {
+            auto expected_size = rows_num - filtered_count;
+            size_t i = 0;
+            Columns new_columns(block.columns());
             for (auto & column : block.getColumns())
             {
-                column = column->filter(filter, filtered_count);
+                new_columns[i++] = column->filter(filter, expected_size);
```

Comment:
Previous implementation didn't work because `block.getColumns()` returns `Columns` by value.
Better to write something like
```cpp
for (auto & column : block)
    column.column = column.column->filter(filter, filtered_count);
```